Feature/notification management

mihaelaaleks opened the pull request:
This PR addresses the following issues: 
1. #1601 
2. #1572

mihaelaaleks committed:
added notification management prototype from feature/notifications

mihaelaaleks committed:
added notification management prototype from feature/notifications

mihaelaaleks committed:
moved notifications to secondary menu

mihaelaaleks committed:
moved notifications to secondary menu

mihaelaaleks committed:
styling changes and fixing inputs in readonly mode

mihaelaaleks committed:
styling changes and fixing inputs in readonly mode

mihaelaaleks committed:
changed add notification btn text

mihaelaaleks committed:
changed add notification btn text

mihaelaaleks committed:
fixed title margins

mihaelaaleks committed:
fixed title margins

mihaelaaleks committed:
added table styling changes

mihaelaaleks committed:
added table styling changes

mihaelaaleks committed:
styling changes to filtering

mihaelaaleks committed:
styling changes to filtering

mihaelaaleks committed:
added rudimentary target resolution

mihaelaaleks committed:
added rudimentary target resolution

mihaelaaleks committed:
fixed target resolution loading bug

mihaelaaleks committed:
fixed target resolution loading bug

mihaelaaleks committed:
update page after realm change

mihaelaaleks committed:
update page after realm change

mihaelaaleks committed:
migration for adding realm column to notification table

mihaelaaleks committed:
migration for adding realm column to notification table

mihaelaaleks committed:
realm column addition to SentNotification model

mihaelaaleks committed:
realm column addition to SentNotification model

mihaelaaleks committed:
fixed asset tree displaying over other form elements

mihaelaaleks committed:
fixed asset tree displaying over other form elements

mihaelaaleks committed:
translated target type list

mihaelaaleks committed:
translated target type list

mihaelaaleks committed:
clean up notification form and unify state management

mihaelaaleks committed:
clean up notification form and unify state management

mihaelaaleks committed:
Merge branch 'master' into feature/notification-management

mihaelaaleks committed:
fixed build fail

mihaelaaleks committed:
fixed build fail

mihaelaaleks committed:
add realm support to notificatiosn for improved filtering

mihaelaaleks committed:
process criteria updated for notification realm ownership

mihaelaaleks committed:
Merge branch 'master' of https://github.com/openremote/openremote into feature/notification-management

mihaelaaleks committed:
updated origin with small changes

mihaelaaleks committed:
moved migration to correct directory

mihaelaaleks committed:
set default realm for INTERNAL notifications to prevent transaction rollback

mihaelaaleks committed:
Merge branch 'master' into feature/notification-management

DonWillems commented:
Hi @mihaelaaleks , do you have a status update?

mihaelaaleks commented:
@DonWillems  Apologies for the overdue update 😅 I have some last tests to add and I'll push my changes for review by tomorrow.

mihaelaaleks committed:
fixed realm ownership issues

mihaelaaleks committed:
added endpoints for realm ownership

mihaelaaleks committed:
cleaned up notification form

mihaelaaleks committed:
cleaned up console logs and comments from notifications page

mihaelaaleks committed:
added separate endpoints for calling notifications with realm ownership

mihaelaaleks committed:
Merge branch 'feature/notification-management' of https://github.com/openremote/openremote into feature/notification-management

mihaelaaleks committed:
Merge branch 'master' into feature/notification-management

DonWillems committed:
Merge branch 'master' into feature/notification-management

DonWillems commented:
Hi Michaela, it took a while for me to find the time, but here are some points that stood out to me:
- [ ] As you mentioned the deliveredOn and acknowledgedOn timestamps aren't set correctly by the apps. I created a separate bug for this: #1819 . If I use the endpoints manually, the table does update the status correctly.
- [ ] Asset tree is awkward in the dialog, as it doesn't fit with many assets (and otherwise looks off too).
![assettreeInDialog](https://github.com/user-attachments/assets/7b60e1bd-3d4e-44aa-8582-da8a197ff95f)
I understand why you chose this over opening it as a dialog. I think we should do it like it is in rules: first select asset type, then a (multi)select of the assets of that type. With at the top of that list the option to select 'All of this type'.
- [ ] Service users shouldn't be shown in the users list
- [ ] You can't multiselect users, which will make sending notifications tedious. Ideally you would also want to be able to select all users that have a certain realm role or role.
- [ ] If notification is sent to user, it still shows as target: asset. I assume there is no way to read this from the notification in the db. I think we should always show the linked user(s) of the target asset to give the needed information. The downside is that if the linked user is changed, the info won't be correct historically. But let's accept that for now.
- [ ] Target is clickable, underline the target so that it is clear that is individually clickable as opposed to the click anywhere else on the row.
- [ ] When creating a notification with an asset as target that does not have a linked user, the notification correctly can't be sent. However, the notification says: "Failed to create notification:" with no explanation.
- [ ] When creating a notification but not filling in all required fields (like title and content) you can press the create button, but you will get an error. The Create button should only be enabled if you filled in all the required fields.
- [ ] Every time I tab back to the table page, its recollecting all the rows?
- [ ] Something weird is going on with sorting on sent date time:
![image](https://github.com/user-attachments/assets/da6c7aac-2acd-4bdd-9da3-998bbe5344dc)
- [ ] realm field stays invalid even if a realm is set:
![Target field stays invalid](https://github.com/user-attachments/assets/09dcdd74-bb76-4587-8090-b99a2593b75b)
- [ ] The rule id that sent the notification is not available? Should we add that?
- [ ] Remove the console messages
- [ ] What does priority do when creating a notification?
- [ ] Why are you doing a call per row? Is it to get the name and icon of the target asset? Find out if there is a better way, if there isn't, maybe we need to leave the name out?

Nice to have:
- [ ] Add a search that matches anything in the 'title', asset id, and 'linked user'. Filter out the rows that match
- [ ] Is there a way to combine notifications that are sent from one rule trigger/from one manual action? If it is sent to 100 consoles at once, you will have 100 entries, would be nice to have that as 1 entry; or an entry per status.

Turned out to be quite a list. If there is anything unclear, feel free to give me a call. Also, if you need some help from the developers don't hesitate to contact them.

DonWillems committed:
Merge branch 'master' into feature/notification-management

DonWillems committed:
Merge branch 'master' into feature/notification-management

DonWillems commented:
Some more notes, will clean up later:

- [ ] Make sure the non-'dev-only' endpoint is used
- [ ] Replace DeliveredOn column with Linked Users column. In practice the table only shows the assets the messages are sent to (needs triple checking). This means the Target is always an asset, even if the user or realm is targeted in the message creation. The assets (consoles) a message can be sent to always have a linked user. It would be helpful to also see that linked user immediately.
- [ ] Make the target field in the dialog of an existing form nicer by using the same layout as in the table cell. However, similar as above, probably all existing notifications will only have assets as target. So consider to replace "Target Type" and "Target" input fields with the icon+text combination for the asset and the user.

DonWillems committed:
- Adjust layout and styling of notification form
- Make it possible to set multiple targets
- Add translations
- Make target in row click more visible

DonWillems committed:
Merge remote-tracking branch 'origin/feature/notification-management' into feature/notification-management

DonWillems committed:
Merge branch 'master' into feature/notification-management

DonWillems committed:
Merge branch 'master' into feature/notification-management

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openremote/openremote?pullRequest=1615) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/openremote/openremote?pullRequest=1615) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: DonWillems<br/>:x: mihaelaaleks<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openremote/openremote?pullRequest=1615) it.</sub>

DonWillems committed:
Merge branch 'master' into feature/notification-management

MartinaeyNL committed:
Merge branch 'master' into feature/notification-management
